### PR TITLE
check-tdx-host.sh: refine check as feedback

### PR DESCRIPTION
**Pull request by intelzhongjie**

### CODE REVIEW

File: utils/check-tdx-host.sh [hunk 1]
```diff
@@ -36,12 +36,41 @@ print_title() {
...
 }
 
+SUPPORTED_OS[0]="Ubuntu 22.04.1 LTS"
+SUPPORTED_OS[1]="Red Hat Enterprise Linux 8.7 (Ootpa)"
+
+#
+# Print the supported OSes.
+#
+print_supported_os_list() {
+    print_guide "Currently we support these OSes:"
```
Comment (eadamsintel) [inline utils/check-tdx-host.sh:46]: I would rephrase this to  "The following OSs are supported:"  Every place It says OSes it should be OSs. I had to look it up and found spirited debates about this but the more common usage is OSs to be plural.

File: utils/check-tdx-host.sh [hunk 2]
```diff
@@ -86,12 +115,16 @@ check_cmd() {
...
+    os_name=$(grep PRETTY_NAME /etc/os-release | sed "s/PRETTY_NAME=\"\(.*\)\"/\1/g")
+    retval=$(is_os_supported "$os_name")
+    [[ "$retval" == "true" ]] && result="OK" || result="FAIL"
+    local action="Check OS: The distro and version are correct (mandatory)"
+    local reason="Your OS distro is not supported yet."
+    report_result "$result" "$action" "$reason" mandatory
+    print_guide "Your current OS is ${os_name}."
+    print_supported_os_list
+    print_guide "There is no guarantee to other OS distro."
     print_guide "Details can be found in Whitepaper: Linux* Stacks for Intel® Trust Domain Extension"
```
Comment (eadamsintel) [inline utils/check-tdx-host.sh:128]: "Details can be found in the Whitepaper:"

File: utils/check-tdx-host.sh [hunk 3]
```diff
@@ -158,24 +179,29 @@ check_bios_tme_bypass() {
     local reason="The bit 31 of MSR 0x982 should be 1"
     local retval
     retval=$(sudo rdmsr -f 31:31 0x982)
-    [[ "$retval" == 1 ]] && result="OK" || result="FAIL"
+    [[ "$retval" == 1 ]] && result="OK" || result="WARNING"
     report_result "$result" "$action" "$reason" optional
+    print_guide "It's better to enable TME Bypass for traditional non-confidential workload."
     print_guide "Details can be found in Intel SDM: Vol. 4 Model Specific Registers (MSRs)"
```
Comment (eadamsintel) [inline utils/check-tdx-host.sh:185]: "Details can be found in the Intel"

File: utils/check-tdx-host.sh [hunk 4]
```diff
@@ -158,24 +179,29 @@ check_bios_tme_bypass() {
     local reason="The bit 31 of MSR 0x982 should be 1"
     local retval
     retval=$(sudo rdmsr -f 31:31 0x982)
-    [[ "$retval" == 1 ]] && result="OK" || result="FAIL"
+    [[ "$retval" == 1 ]] && result="OK" || result="WARNING"
     report_result "$result" "$action" "$reason" optional
+    print_guide "It's better to enable TME Bypass for traditional non-confidential workload."
```
Comment (eadamsintel) [inline utils/check-tdx-host.sh:184]: "for traditional non-confidential workloads"  The last word should be plural.

File: utils/check-tdx-host.sh [hunk 5]
```diff
@@ -158,24 +179,29 @@ check_bios_tme_bypass() {
...
+    local action="Check BIOS: TME-MT/TME-MK (mandatory & manually)"
+    local reason="The bit 1 of MSR 0x982 should be 1"
+    local retval
+    retval=$(sudo rdmsr -f 1:1 0x982)
+    [[ "$retval" == 1 ]] && result="OK" || result="FAIL"
+    report_result "$result" "$action" "$reason" mandatory
     print_guide "Please check your BIOS settings:"
     print_guide "    Socket Configuration -> Processor Configuration -> TME, TME-MT, TDX"
     print_guide "        Total Memory Encryption Multi-Tenant (TME-MT) should be Enable"
     print_guide "(Different BIOS may need different setting path)"
```
Comment (eadamsintel) [inline utils/check-tdx-host.sh:203]: I would reword this to  "A different BIOS might have a different path for this setting"

**Comment by eadamsintel** [comment]
Is it not possible to get the TDX module version through this script? If not can you give instructions on how someone can check manually? Same comment for the SEAM loader.  Also, the same color yellow is used for warning and TBD but those convey different meanings.  Maybe the "Warning" should be orange?

File: utils/check-tdx-host.sh [hunk 6]
```diff
@@ -110,25 +143,13 @@ check_tdx_module() {
...
+    print_guide "Please check your TDX Module info to see if that's what you need."
     print_guide "Details can be found in Whitepaper: Linux* Stacks for Intel® Trust Domain Extension"
     print_url $URL_TDX_LINUX_WHITE_PAPER
     printf "\n"
 }
 
-#
-# TDX only support 1LM mode
-#
-check_bios_memory_map() {
```
Comment (ruomengh) [inline utils/check-tdx-host.sh:121]: This is still needed in some BIOS. I would suggest to keep this part and add a note "Please skip this setting if it doesn't exist in your BIOS menu"

File: utils/check-tdx-host.sh [hunk 7]
```diff
@@ -158,24 +179,29 @@ check_bios_tme_bypass() {
...
 
 #
-# Check TME-MT setting in BIOS
+# Check TME-MT/TME-MK setting in BIOS
 #
 check_bios_tme_mt() {
-    local action="Check BIOS: TME-MT (mandatory & manually)"
-    local reason=""
-    report_result TBD "$action" "$reason" mandatory manual
+    local action="Check BIOS: TME-MT/TME-MK (mandatory & manually)"
```
Comment (ruomengh) [inline utils/check-tdx-host.sh:194]: The BIOS knob is TME-MT. Which BIOS knob does TME-MK refer to?

File: utils/check-tdx-host.sh (hunk 1 above)
Comment (intelzhongjie) [reply utils/check-tdx-host.sh]: Updated. Please review the latest version.

File: utils/check-tdx-host.sh (hunk 2 above)
Comment (intelzhongjie) [reply utils/check-tdx-host.sh]: Updated. Please review the latest version.

File: utils/check-tdx-host.sh (hunk 3 above)
Comment (intelzhongjie) [reply utils/check-tdx-host.sh]: Updated. Please review the latest version.

File: utils/check-tdx-host.sh (hunk 4 above)
Comment (intelzhongjie) [reply utils/check-tdx-host.sh]: Updated. Please review the latest version.

File: utils/check-tdx-host.sh (hunk 5 above)
Comment (intelzhongjie) [reply utils/check-tdx-host.sh]: Updated. Please review the latest version.

**Comment by intelzhongjie** [comment]
> Is it not possible to get the TDX module version through this script? If not can you give instructions on how someone can check manually? Same comment for the SEAM loader. Also, the same color yellow is used for warning and TBD but those convey different meanings. Maybe the "Warning" should be orange?

**Comment by intelzhongjie** [comment]
> Is it not possible to get the TDX module version through this script? If not can you give instructions on how someone can check manually? Same comment for the SEAM loader. Also, the same color yellow is used for warning and TBD but those convey different meanings. Maybe the "Warning" should be orange?

Regarding the TDX module version, we need some clarifications: the current script does print the TDX module version in its output. But users do need to check the output version manually to determine whether the version is as the expectation or not. e.g. User A may need some version for 1.0. User B may need some version for 1.5. Which version to use is a decision by users.
Similar situation for SEAM loader option.

Regarding to the text color for "Warning", I changed that to magenta in the latest version because I tried some "orange" text but haven't found a proper one that can be easily distinguished from red and yellow. Please review this and let me know your thoughts. Thanks!

File: utils/check-tdx-host.sh (hunk 6 above)
Comment (intelzhongjie) [reply utils/check-tdx-host.sh]: Updated. Please review the latest version.

File: utils/check-tdx-host.sh (hunk 7 above)
Comment (intelzhongjie) [reply utils/check-tdx-host.sh:215]: I saw SDM use TME-MK. Please let me find out the connection between these 2 terms and update later. Thanks!

**Comment by bktan8** [comment]
I find the current format and output confusing to understand.  For example, "mandatory & manually".  I recommend splitting the mandatory and optional sections.  I realize this is more effort, but it will be much more user friendly and easier to comprehend.  Here's my example of what would be better:

```
   *** TDX Host Check ***
 
Mandatory Features & Settings
------------------------------------------
1) Check Linux distro and version.........FAIL
   Result: Ubuntu 22.04.3 LTS.  This is not supported.  
  
  Note: 
   Supported OSes:
        Ubuntu 22.04.1 LTS
        Red Hat Enterprise Linux 8.7 (Ootpa)
    Details can be found in Whitepaper: Linux* Stacks for Intel® Trust Domain Extension
    Link: https://www.intel.com/content/www/us/en/content-details/779108/whitepaper-linux-stacks-for-intel-trust-domain-extension-1-0.html
 
1) Check Linux distro and version.........SUCCESS
  Result: Ubuntu 22.04.1 LTS

2) Check TDX Module version ..... FAIL
    Result: Unable to check through this script. Please check manually.

    Note: 
    Different releases could require different versions of TDX module.
    Please check your TDX Module info to see if that's what you need.
    Details can be found in Whitepaper: Linux* Stacks for Intel® Trust Domain Extension
    Link: https://www.intel.com/content/www/us/en/content-details/779108/whitepaper-linux-stacks-for-intel-trust-domain-extension-1-0.html

2) Check TDX Module version ..... SUCCESS
Result: 0x00000000 build_date: 20220420 build_num: 0x00000156 major_version: 0x00000001 minor_version: 0x00000000 status: initialized vendor_id: 0x000                                            08086
    
Note:
    Different releases could require different versions of TDX module.
    Please check your TDX Module info to see if that's what you need.
    Details can be found in Whitepaper: Linux* Stacks for Intel® Trust Domain Extension
    Link: https://www.intel.com/content/www/us/en/content-details/779108/whitepaper-linux-stacks-for-intel-trust-domain-extension-1-0.html

3) Check BIOS setting: Total Memory Encryption (TME)................................ FAIL
    Result: Disabled

   Note: 
    Details can be found in Intel SDM: Vol. 4 Model Specific Registers (MSRs)
    Link: https://www.intel.com/content/www/us/en/developer/articles/technical/intel-sdm.html
 
3) Check BIOS setting: Total Memory Encryption (TME) ................................ SUCCESS
    Result: Enabled

   Note:
    Details can be found in Intel SDM: Vol. 4 Model Specific Registers (MSRs)
    Link: https://www.intel.com/content/www/us/en/developer/articles/technical/intel-sdm.html

4) Check BIOS setting: Total Memory Encryption Multi-Tenant (TME-MT) ..................... FAIL
   Result: Disabled

  Note: 
    Please enable this setting in your BIOS:
        Socket Configuration -> Processor Configuration -> TME, TME-MT, TDX
            Total Memory Encryption Multi-Tenant (TME-MT) -> Enable
    
(Different BIOS may need different setting path)
    Link: https://www.intel.com/content/www/us/en/developer/articles/technical/intel-sdm.html

4) Check BIOS setting: Total Memory Encryption Multi-Tenant (TME-MT) ..................... SUCCESS
   Result: Enabled

5) Check BIOS setting: Trusted Domain Extension (TDX)  ................................ FAIL
    Result: Disabled

  Note: 
    Please enable this setting in your BIOS:
        Socket Configuration -> Processor Configuration -> TME, TME-MT, TDX
            Trusted Domain Extension (TDX) -> Enable

5) Check BIOS setting: Trusted Domain Extension (TDX)  ................................ SUCCESS
    Result: Enabled

6) Check BIOS setting: SW Guard Extensions (SGX)  ................................ FAIL
    Result: Disabled

  Note: 
    Please enable this setting in your BIOS:
        Socket Configuration -> Processor Configuration -> Sofware Guard Extension ->
        SW Guard Extensions (SGX) -> Enable

6) Check BIOS setting: SW Guard Extensions (SGX)  ................................ SUCCESS
    Result: Enabled
 
7) Check BIOS setting: TDX Key Split != 0 ........................... 

8) Check BIOS setting: SGX registration server........... 
 
 
Optional Features & Settings
------------------------------------------
9) Check BIOS setting: TME Bypass.......................... 
 
10) Check BIOS: SEAM Loader  ......................... 
 
 
    We highly recommend you to check the output info above seriously and
    follow the corresponding guide carefully because that's the fastest way
    for your troubleshooting. Attention to the text in red or yellow.
    That being said, if you can't resolve your problem after checking all
    the items above, neither in program nor manually, then you can contact
    maintainers of http://github.com/intel/tdx-tool for support with your
    output of this script.

```

**Comment by eadamsintel** [comment]
> Regarding to the text color for "Warning", I changed that to magenta in the latest version because I tried some "orange" text but haven't found a proper one that can be easily distinguished from red and yellow. Please review this and let me know your thoughts. Thanks!

I like the magenta better. I also agree with Bun to split the mandatory and manually. I am trying to think of a better way to do it.  Instead of mandatory maybe use the word "Required" instead since "optional" is used in other places.

**Comment by intelzhongjie** [comment]
OK, I'll update the change with the suggestions from @bktan8 and @eadamsintel later.

**Comment by intelzhongjie** [comment]
Updated per feedback. The remaining issue is about the MT vs. MK. I'll give update to that later. Thanks!

**Comment by intelzhongjie** [comment]
FYI. Corresponding PR for other branches:
tdx-1.5 https://github.com/intel/tdx-tools/pull/472
tdx-1.0 https://github.com/intel/tdx-tools/pull/473

File: utils/check-tdx-host.sh [hunk 8]
```diff
@@ -101,16 +146,18 @@ check_os() {
...
 #
 check_tdx_module() {
-    local action="Check TDX Module: The version is expected (mandatory & manually)"
+    local action="Check TDX Module: The version is expected (required & manually)"
     local reason=""
-    report_result TBD "$action" "$reason" mandatory manual
+    report_result TBD "$action" "$reason" required manual
     local tdx_module_info
     # shellcheck disable=SC2012
     tdx_module_info=$(ls /sys/firmware/tdx/tdx_module | while read -r tdxattr; \
```
Comment (eadamsintel) [inline utils/check-tdx-host.sh:154]: On my system I did not have the TDX module. It showed a yellow "TBD" but below it the message said  "ls: cannot access /sys/firmware/tdx/tdx_module"   In this scenario I think it should be a red "Fail" since the module does not exist. In the other case yellow is good to use.

Comment (intelzhongjie) [reply utils/check-tdx-host.sh]: Thanks! This is updated. Please review the latest version.

**Comment by eadamsintel** [comment]
It looks better in red on my system without a TDX module. I just got back my other system that has the module installed and tried the script out. On that system it had the yellow TBD and had the following. I still don't have any idea how to verify the TDX module version and that it is the correct version. Is there a specific chapter in the white paper that explains how to do this? When I look at `/sys/firmware/tdx/tdx_module/status` it just says unknown the same as the script. I think at the very least in the "TBD" case it should state something like "TDX module found at /sys/firmware/tdx/tdx_module/status  status: Unknown"   However, we need to provide more information on how to ensure they have the right one.  Is that possible in any way?  I assume that the only thing that matters at this stage is the 1.5 TDX module is used with the 1.5 stack and 1.0 module is used with the 1.0 stack. Is it possible to do a sha hash of the module and compare against released versions on the GitHub to determine which release it belongs to    ex:  TDX module is the v1.5 2023ww33 release.   Thoughts?

```
Check TDX Module: The version is expected (required & manually) ...... TBD
    Reason: Unable to check in program. Please check manually.
    Your TDX Module info. status: unknown
    Different releases could require different versions of TDX module.
    Please check your TDX Module info to see if that's what you need.
    Details can be found in the Whitepaper: Linux* Stacks for Intel® Trust Domain Extension
    Link: https://www.intel.com/content/www/us/en/content-details/779108/whitepaper-linux-stacks-for-intel-trust-domain-extension-1-0.html
```

**Comment by eadamsintel** [comment]
An easier way to do this would be to add version information to the TDX module that you could probe with this script.

**Comment by intelzhongjie** [comment]
> It looks better in red on my system without a TDX module. I just got back my other system that has the module installed and tried the script out. On that system it had the yellow TBD and had the following. I still don't have any idea how to verify the TDX module version and that it is the correct version. Is there a specific chapter in the white paper that explains how to do this? When I look at `/sys/firmware/tdx/tdx_module/status` it just says unknown the same as the script. I think at the very least in the "TBD" case it should state something like "TDX module found at /sys/firmware/tdx/tdx_module/status status: Unknown" However, we need to provide more information on how to ensure they have the right one. Is that possible in any way? I assume that the only thing that matters at this stage is the 1.5 TDX module is used with the 1.5 stack and 1.0 module is used with the 1.0 stack. Is it possible to do a sha hash of the module and compare against released versions on the GitHub to determine which release it belongs to ex: TDX module is the v1.5 2023ww33 release. Thoughts?
> 
> ```
> Check TDX Module: The version is expected (required & manually) ...... TBD
>     Reason: Unable to check in program. Please check manually.
>     Your TDX Module info. status: unknown
>     Different releases could require different versions of TDX module.
>     Please check your TDX Module info to see if that's what you need.
>     Details can be found in the Whitepaper: Linux* Stacks for Intel® Trust Domain Extension
>     Link: https://www.intel.com/content/www/us/en/content-details/779108/whitepaper-linux-stacks-for-intel-trust-domain-extension-1-0.html
> ```

Thanks for your feedback! Please let me clarify why it's "TBD" here furtherly as below.
Besides the "1.0 vs. 1.5" difference, different vendors may use different versions for different hardware boards. Then given one specific version, how can we use this script to determine if that version is the correct version for the hardware on which it's running?

I think this script does some basic check on the configurations. More complex check like "software versions matched hardware versions and integrity is OK" could be done by attestation/measurement, correct?

Please let me know if I misunderstood or missed your point.

**Comment by eadamsintel** [comment]
> I think this script does some basic check on the configurations. More complex check like "software versions matched hardware versions and integrity is OK" could be done by attestation/measurement, correct?
> 
> Please let me know if I misunderstood or missed your point.

I think the issue is that `/sys/firmware/tdx/tdx_module/status` just says Unknown and does not give the version information. The script says to look at the white paper when it is TBD, but that does not tell you how to check what version of the module is installed.

**Comment by intelzhongjie** [comment]
> > I think this script does some basic check on the configurations. More complex check like "software versions matched hardware versions and integrity is OK" could be done by attestation/measurement, correct?
> > Please let me know if I misunderstood or missed your point.
> 
> I think the issue is that `/sys/firmware/tdx/tdx_module/status` just says Unknown and does not give the version information. The script says to look at the white paper when it is TBD, but that does not tell you how to check what version of the module is installed.

That "unknown" is a problem. It means "TDX module hasn't been detected and initialized". Normally, "/sys/firmware/tdx/tdx_module/status" should be "initialized" when TDX module succeed at loading. Please let me double check and get you back later.

**Comment by eadamsintel** [comment]
That "unknown" is because my system was just returned to me after a BIOS upgrade and all the TDX settings were disabled. I re-enabled everything and now get "error" when I checked the TDX module.  I was setting up the TDX 1.5 stack a few weeks ago on this system when I decided to update the BIOS. I think my module is in a bad state. I enabled the SEAM loader in the BIOS to load the module from BIOS instead of from the OS and it now says "Initialized"

```
Check TDX Module: The version is expected (required & manually) ...... TBD
    Reason: Unable to check in program. Please check manually.
    Your TDX Module info. attributes: 0x00000000 build_date: 20220420 build_num: 0x00000156 major_version: 0x00000001 minor_version: 0x00000000 status: initialized vendor_id: 0x00008086
    Different releases could require different versions of TDX module.
    Please check your TDX Module info to see if that's what you need.
    Details can be found in the Whitepaper: Linux* Stacks for Intel® Trust Domain Extension
    Link: https://www.intel.com/content/www/us/en/content-details/779108/whitepaper-linux-stacks-for-intel-trust-domain-extension-1-0.html
```

It still isn't clear to me 100% if I have the correct module or not. I think linking to a table of supported versions in a release along with how to identify which release you are using would help a lot. I don't now if that is possible. 

I noticed a discrepancy between the script and the WP.  The WP says that "TME bypass" should be auto and that is what is set in my BIOS, but the script gives a warning and says it should be "enabled"

**Comment by intelzhongjie** [comment]
We will merge the latest change for this through:
tdx-1.5 https://github.com/intel/tdx-tools/pull/472
tdx-1.0 https://github.com/intel/tdx-tools/pull/473

And we will discuss the tdx module check separately.
Thanks!

**Comment by intelzhongjie** [comment]
We will discuss the tdx module check later in a different thread.
